Source/plrmsg.cpp

```diff
@@ -9,32 +9,45 @@
 
 #include <fmt/format.h>
 
-#include "DiabloUI/ui_flags.hpp"
 #include "control.h"
 #include "engine/render/text_render.hpp"
 #include "inv.h"
 #include "utils/language.h"
-#include "utils/stdcompat/string_view.hpp"
 #include "utils/utf8.hpp"
 
 namespace devilution {
 
 namespace {
 
-#define PMSG_COUNT 8
+struct PlayerMessage {
+	/** Time message was recived */
+	Uint32 time;
+	/** The default text color */
+	UiFlags style;
+	/** The text message to display on screen */
+	std::string text;
+	/** Portion of text that should be rendered in gold */
+	std::string from;
+	/** The line height of the text */
+	int lineHeight;
+};
+
+std::array<PlayerMessage, 8> Messages;
+
+int CountLinesOfText(string_view text)
+{
+	return 1 + std::count(text.begin(), text.end(), '\n');
+}
 
-uint8_t plr_msg_slot;
-_plrmsg plr_msgs[PMSG_COUNT];
+PlayerMessage &GetNextMessage()
+{
+	PlayerMessage &front = Messages.front();
 
-/** Maps from player_num to text color, as used in chat messages. */
-const UiFlags TextColorFromPlayerId[MAX_PLRS + 1] = { UiFlags::ColorWhite, UiFlags::ColorWhite, UiFlags::ColorWhite, UiFlags::ColorWhite, UiFlags::ColorWhitegold };
+	std::move_backward(&front, &Messages.back(), Messages.end());
 
-void PrintChatMessage(const Surface &out, int x, int y, int width, char *textPtr, UiFlags style)
-{
-	const size_t length = strlen(textPtr);
-	std::replace(textPtr, textPtr + length, '\n', ' ');
-	const string_view text { textPtr, length };
-	DrawString(out, WordWrapString(text, width), { { x, y }, { width, 0 } }, style, 1, 18);
+	front = {};
+
+	return front;

```

Comment:
If you're willing to avoid mixing the iterator/pointer types in the previous call to `move_backwards` then I think it's clearer to move the declaration of `front` here.
```suggestion
	PlayerMessage &front = Messages.front();
	front = {};
	return front;
```

`operator=` returns a self reference as well, if method chaining is acceptable the temporary front variable could be completely removed.
```suggestion
	return Messages.front() = {};
```

Reply:
Going to drop the reset as the setters now fill out all values instead.